### ISSUE TITLE
bigip_data_group doesn't support route domain notation in an IP type data group

### ISSUE THREAD

**Issue by writemike** [F5Networks/f5-ansible#1346]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

bigip_data_group Module
##### ANSIBLE VERSION

```
ansible 2.8.0
  config file = /home/f5/Documents/optus-auto/ansible.cfg
  configured module search path = [u'/home/f5/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/f5/.local/lib/python2.7/site-packages/ansible
  executable location = /home/f5/.local/bin/ansible
  python version = 2.7.5 (default, Apr  9 2019, 14:30:50) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]
```

##### PYTHON VERSION

```
Python 2.7.5
```

##### BIGIP VERSION

```
Sys::Version
Main Package
  Product     BIG-IP
  Version     13.1.1
  Build       0.0.4
  Edition     Final
  Date        Fri Jul 20 17:55:49 PDT 2018
```

##### ROLE VERSION

```
Role: f5devcentral.f5ansible
        description: Ansible role that allows use of the F5 Ansible Modules
        active: True
        commit: 7d42617a49944229f5b30c3a69a4b394dd14a615
        commit_message: Merging upstream changes - 2019-05-20T07:01:08Z
        commit_url: https://api.github.com/repos/f5devcentral/ansible-role-f5ansible/git/commits/7d42617a49944229f5b30c3a69a4b394dd14a615
        company: F5 Networks
        created: 2018-10-04T16:14:10.510085Z
        dependencies: []
        download_count: 5033
        forks_count: 0
        galaxy_info:
                author: caphrim007
                company: F5 Networks
                galaxy_tags: ['f5', 'networking', 'bigip', 'bigiq']
                license: GPLv3
                min_ansible_version: 2.7
                platforms: [{'name': 'EL', 'versions': ['all']}, {'name': 'Fedora', 'versions': ['all']}, {'name': 'Ubuntu', 'versions': ['all']}, {'name': 'Debian', 'versions': ['all']}]
                role_name: f5ansible
        github_branch: master
        github_repo: ansible-role-f5ansible
        github_user: f5devcentral
        id: 30341
        imported: 2019-05-20T03:03:59.440544-04:00
        install_date: Tue May 21 09:37:33 2019
        installed_version: v2019.5.10
        is_valid: True
        issue_tracker_url:
        license: GPLv3
        min_ansible_version: 2.8
        modified: 2019-05-20T07:03:59.635888Z
        open_issues_count: 0
        path: (u'/home/f5/.ansible/roles', u'/usr/share/ansible/roles', u'/etc/ansible/roles')
        role_type: ANS
        stargazers_count: 4
        travis_status_url: https://travis-ci.org/f5devcentral/ansible-role-f5ansible.svg?branch=master
```

##### CONFIGURATION

$ cat ansible.cfg
[defaults]
inventory      = ./hosts
$ cat ./hosts
[bigip]
192.168.200.245
##### OS / ENVIRONMENT

```
$ uname -a
Linux controlplane-mypemlab-local 3.10.0-957.12.2.el7.x86_64 #1 SMP Tue May 14 21:24:32 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```
##### SUMMARY

We are using the bigip_data_group to update an existing internal IP Data Group. The address entered needs to contain the route domain notation (%<RD#>) in order for the iRule to work. We have been adding several IP entries with route domain notation through tmsh and the WebGUI with no issues.

```
(tmos)# create ltm data-group internal Data_Group_Peer type ip
(tmos)# modify ltm data-group internal Data_Group_Peer records add { 10.0.0.20%1508/32 { data 7020 } }
(tmos)# list ltm data-group internal Data_Group_Peer
ltm data-group internal Data_Group_Peer {
    records {
        10.0.0.20%1508/32 {
            data 7020
        }
    }
    type ip
}
```

##### STEPS TO REPRODUCE

```
---

- name: Test Data-Group Updates
  hosts: bigip
  connection: local
  gather_facts: no
  vars_files:
   - onboarding-vars.yaml

  roles:
    - role: f5devcentral.f5ansible

  tasks:
    - name: Add Tenant to Data-Group
      bigip_data_group:
        provider: "{{ provider }}"
        name: "Data_Group_Peer"
        internal: yes
        type: ip
        records:
          - key: "10.0.0.4%1508/32"
            value: "7004"
        state: present
        partition: Common
      delegate_to: localhost
```

##### EXPECTED RESULTS

The Data-Group should contain the new entry with the route domain notation.

##### ACTUAL RESULTS

We get an error:

```
$ ansible-playbook dg-update-test.yaml -vvv
ansible-playbook 2.8.0
  config file = /home/f5/Documents/optus-auto/ansible.cfg
  configured module search path = [u'/home/f5/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/f5/.local/lib/python2.7/site-packages/ansible
  executable location = /home/f5/.local/bin/ansible-playbook
  python version = 2.7.5 (default, Apr  9 2019, 14:30:50) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]
Using /home/f5/Documents/optus-auto/ansible.cfg as config file
host_list declined parsing /home/f5/Documents/optus-auto/hosts as it did not pass it's verify_file() method
script declined parsing /home/f5/Documents/optus-auto/hosts as it did not pass it's verify_file() method
auto declined parsing /home/f5/Documents/optus-auto/hosts as it did not pass it's verify_file() method
Parsed /home/f5/Documents/optus-auto/hosts inventory source with ini plugin
Read vars_file 'onboarding-vars.yaml'
Read vars_file 'onboarding-vars.yaml'

PLAYBOOK: dg-update-test.yaml ************************************************************************************************************************************************************************************************************************************
1 plays in dg-update-test.yaml
Read vars_file 'onboarding-vars.yaml'
Read vars_file 'onboarding-vars.yaml'
Read vars_file 'onboarding-vars.yaml'

PLAY [Test Data-Group Updates] ***********************************************************************************************************************************************************************************************************************************
META: ran handlers
Read vars_file 'onboarding-vars.yaml'

TASK [f5devcentral.f5ansible : Check ansible version] ************************************************************************************************************************************************************************************************************
task path: /home/f5/.ansible/roles/f5devcentral.f5ansible/tasks/main.yaml:7
skipping: [192.168.200.245] => {}
Read vars_file 'onboarding-vars.yaml'
Read vars_file 'onboarding-vars.yaml'

TASK [Add Tenant to Data-Group] **********************************************************************************************************************************************************************************************************************************
task path: /home/f5/Documents/optus-auto/dg-update-test.yaml:14
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: f5
<localhost> EXEC /bin/sh -c 'echo ~f5 && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/f5/.ansible/tmp/ansible-tmp-1558673212.11-60959083507252 `" && echo ansible-tmp-1558673212.11-60959083507252="` echo /home/f5/.ansible/tmp/ansible-tmp-1558673212.11-60959083507252 `" ) && sleep 0'
Using module file /home/f5/.ansible/roles/f5devcentral.f5ansible/library/bigip_data_group.py
<localhost> PUT /home/f5/.ansible/tmp/ansible-local-14542VARFra/tmpI_w6eI TO /home/f5/.ansible/tmp/ansible-tmp-1558673212.11-60959083507252/AnsiballZ_bigip_data_group.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/f5/.ansible/tmp/ansible-tmp-1558673212.11-60959083507252/ /home/f5/.ansible/tmp/ansible-tmp-1558673212.11-60959083507252/AnsiballZ_bigip_data_group.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python2 /home/f5/.ansible/tmp/ansible-tmp-1558673212.11-60959083507252/AnsiballZ_bigip_data_group.py && sleep 0'
<localhost> EXEC /bin/sh -c 'rm -f -r /home/f5/.ansible/tmp/ansible-tmp-1558673212.11-60959083507252/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
WARNING: The below traceback may *not* be related to the actual failure.
  File "/tmp/ansible_bigip_data_group_payload_MwdKnI/__main__.py", line 1313, in main
    results = mm.exec_module()
  File "/tmp/ansible_bigip_data_group_payload_MwdKnI/__main__.py", line 1257, in exec_module
    return manager.exec_module()
  File "/tmp/ansible_bigip_data_group_payload_MwdKnI/__main__.py", line 782, in exec_module
    changed = self.present()
  File "/tmp/ansible_bigip_data_group_payload_MwdKnI/__main__.py", line 831, in present
    return self.create()
  File "/tmp/ansible_bigip_data_group_payload_MwdKnI/__main__.py", line 849, in create
    self._set_changed_options()
  File "/tmp/ansible_bigip_data_group_payload_MwdKnI/__main__.py", line 804, in _set_changed_options
    if getattr(self.want, key) is not None:
  File "/tmp/ansible_bigip_data_group_payload_MwdKnI/__main__.py", line 670, in records
    if self.records_src is None:
  File "/tmp/ansible_bigip_data_group_payload_MwdKnI/__main__.py", line 574, in records_src
    self._write_records_to_file(records)
  File "/tmp/ansible_bigip_data_group_payload_MwdKnI/__main__.py", line 582, in _write_records_to_file
    result = encoder.encode(record)
  File "/tmp/ansible_bigip_data_group_payload_MwdKnI/__main__.py", line 350, in encode
    return self.encode_dict(record)
  File "/tmp/ansible_bigip_data_group_payload_MwdKnI/__main__.py", line 356, in encode_dict
    return self.encode_address_from_dict(record)
  File "/tmp/ansible_bigip_data_group_payload_MwdKnI/__main__.py", line 367, in encode_address_from_dict
    "When specifying an 'address' type, the value to the left of the separator must be an IP."

fatal: [192.168.200.245]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "auth_provider": null,
            "delete_data_group_file": null,
            "description": null,
            "external_file_name": null,
            "internal": true,
            "name": "Data_Group_Peer",
            "partition": "Common",
            "password": null,
            "provider": {
                "auth_provider": null,
                "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
                "proxy_to": null,
                "server": "192.168.200.245",
                "server_port": 443,
                "ssh_keyfile": null,
                "timeout": null,
                "transport": "rest",
                "user": "admin",
                "validate_certs": false
            },
            "records": [
                {
                    "key": "10.0.0.4%1508/32",
                    "value": "7004"
                }
            ],
            "records_src": null,
            "separator": ":=",
            "server": null,
            "server_port": null,
            "state": "present",
            "transport": null,
            "type": "ip",
            "user": null,
            "validate_certs": null
        }
    },
    "msg": "When specifying an 'address' type, the value to the left of the separator must be an IP."
}

PLAY RECAP *******************************************************************************************************************************************************************************************************************************************************
192.168.200.245            : ok=0    changed=0    unreachable=0    failed=1    skipped=1    rescued=0    ignored=0

```

**Comment by focrensh** [comment]
Tracking under FMFA-265

**Comment by wojtek0806** [comment]
There was a fix for a DG created in separate partition with a different default RD: #1210, to avoid further issues, we will introduce route_domain parameter for DG ip internal types

**Comment by wojtek0806** [comment]
this issue is an extension of #1210, the fix for #1210 was not sufficient, the docs state now that you must always specify route domain if the route domain should be in use regardless if the partition where the DG resides has that route domain as default. This is true only for route domains other than the system default 0. Fix will follow shortly for this
